### DISCUSSION TITLE
django.db.utils.OperationalError: connection to server at "172.17.0.1", port 5432 failed: Connection refused         Is the server running on that host and accepting TCP/IP connections?

### DISCUSSION THREAD

**Discussion by Nebelfuchs** [pretix/pretix#3184] [Q&A] [answered]
Hi there,

I am trying to install pretix on a Debian 11 via this guide https://docs.pretix.eu/en/latest/admin/installation/docker_smallscale.html

I do realize that the error message is telling pretix can't connect to the database but I don't see why this is happening.

 Postgres DB is running locally on the Debian 11 VM

> ERROR 2023-03-27 12:34:14,795 django.request log Internal Server Error: /
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.11/site-packages/django/db/backends/base/base.py", line 219, in ensure_connection
>     self.connect()
>   File "/usr/local/lib/python3.11/site-packages/django/utils/asyncio.py", line 33, in inner
>     return func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/backends/base/base.py", line 200, in connect
>     self.connection = self.get_new_connection(conn_params)
>                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/utils/asyncio.py", line 33, in inner
>     return func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/backends/postgresql/base.py", line 187, in get_new_connection
>     connection = Database.connect(**conn_params)
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/psycopg2/__init__.py", line 122, in connect
>     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> psycopg2.OperationalError: connection to server at "172.17.0.1", port 5432 failed: Connection refused
>         Is the server running on that host and accepting TCP/IP connections?
> 
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.11/site-packages/django/core/handlers/exception.py", line 47, in inner
>     response = get_response(request)
>                ^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/utils/deprecation.py", line 116, in __call__
>     response = self.process_request(request)
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/pretix/src/pretix/multidomain/middlewares.py", line 85, in process_request
>     kd = KnownDomain.objects.select_related('organizer', 'event').get(domainname=domain)  # noqa
>          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 431, in get
>     num = len(clone)
>           ^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 262, in __len__
>     self._fetch_all()
>   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 1324, in _fetch_all
>     self._result_cache = list(self._iterable_class(self))
>                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 51, in __iter__
>     results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/compiler.py", line 1173, in execute_sql
>     cursor = self.connection.cursor()
>              ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/utils/asyncio.py", line 33, in inner
>     return func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/backends/base/base.py", line 259, in cursor
>     return self._cursor()
>            ^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/backends/base/base.py", line 235, in _cursor
>     self.ensure_connection()
>   File "/usr/local/lib/python3.11/site-packages/django/utils/asyncio.py", line 33, in inner
>     return func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/backends/base/base.py", line 218, in ensure_connection
>     with self.wrap_database_errors:
>   File "/usr/local/lib/python3.11/site-packages/django/db/utils.py", line 90, in __exit__
>     raise dj_exc_value.with_traceback(traceback) from exc_value
>   File "/usr/local/lib/python3.11/site-packages/django/db/backends/base/base.py", line 219, in ensure_connection
>     self.connect()
>   File "/usr/local/lib/python3.11/site-packages/django/utils/asyncio.py", line 33, in inner
>     return func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/backends/base/base.py", line 200, in connect
>     self.connection = self.get_new_connection(conn_params)
>                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/utils/asyncio.py", line 33, in inner
>     return func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/django/db/backends/postgresql/base.py", line 187, in get_new_connection
>     connection = Database.connect(**conn_params)
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/psycopg2/__init__.py", line 122, in connect
>     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> django.db.utils.OperationalError: connection to server at "172.17.0.1", port 5432 failed: Connection refused
>         Is the server running on that host and accepting TCP/IP connections?

is the complete error message in pretix log

**postgresql.conf:**

> # - Connection Settings -
> 
> listen_addresses = 'localhost,172.17.0.1'               # what IP address(es) to listen on;
>                                         # comma-separated list of addresses;
>                                         # defaults to 'localhost'; use '*' for all
>                                         # (change requires restart)
> port = 5432                             # (change requires restart)
> max_connections = 100                   # (change requires restart)
> #superuser_reserved_connections = 3     # (change requires restart)
> unix_socket_directories = '/var/run/postgresql' # comma-separated list of directories
>                                         # (change requires restart)
> #unix_socket_group = ''                 # (change requires restart)
> #unix_socket_permissions = 0777         # begin with 0 to use octal notation
>                                         # (change requires restart)
> #bonjour = off                          # advertise server via Bonjour
>                                         # (change requires restart)
> #bonjour_name = ''                      # defaults to the computer name
>                                         # (change requires restart)
> 
> # - TCP settings -
> # see "man tcp" for details
> 
> #tcp_keepalives_idle = 0                # TCP_KEEPIDLE, in seconds;
>                                         # 0 selects the system default
> #tcp_keepalives_interval = 0            # TCP_KEEPINTVL, in seconds;
>                                         # 0 selects the system default
> #tcp_keepalives_count = 0               # TCP_KEEPCNT;
>                                         # 0 selects the system default
> #tcp_user_timeout = 0                   # TCP_USER_TIMEOUT, in milliseconds;
>                                         # 0 selects the system default
> 
> #client_connection_check_interval = 0   # time between checks for client
>                                         # disconnection while running queries;
>                                         # 0 for never

**pg_hba.conf:**

> # PostgreSQL Client Authentication Configuration File
> # ===================================================
> #
> # Refer to the "Client Authentication" section in the PostgreSQL
> # documentation for a complete description of this file.  A short
> # synopsis follows.
> #
> # This file controls: which hosts are allowed to connect, how clients
> # are authenticated, which PostgreSQL user names they can use, which
> # databases they can access.  Records take one of these forms:
> #
> # local         DATABASE  USER  METHOD  [OPTIONS]
> # host          DATABASE  USER  ADDRESS  METHOD  [OPTIONS]
> # hostssl       DATABASE  USER  ADDRESS  METHOD  [OPTIONS]
> # hostnossl     DATABASE  USER  ADDRESS  METHOD  [OPTIONS]
> # hostgssenc    DATABASE  USER  ADDRESS  METHOD  [OPTIONS]
> # hostnogssenc  DATABASE  USER  ADDRESS  METHOD  [OPTIONS]
> #
> # (The uppercase items must be replaced by actual values.)
> #
> # The first field is the connection type:
> # - "local" is a Unix-domain socket
> # - "host" is a TCP/IP socket (encrypted or not)
> # - "hostssl" is a TCP/IP socket that is SSL-encrypted
> # - "hostnossl" is a TCP/IP socket that is not SSL-encrypted
> # - "hostgssenc" is a TCP/IP socket that is GSSAPI-encrypted
> # - "hostnogssenc" is a TCP/IP socket that is not GSSAPI-encrypted
> #
> # DATABASE can be "all", "sameuser", "samerole", "replication", a
> # database name, or a comma-separated list thereof. The "all"
> # keyword does not match "replication". Access to replication
> # must be enabled in a separate record (see example below).
> #
> # USER can be "all", a user name, a group name prefixed with "+", or a
> # comma-separated list thereof.  In both the DATABASE and USER fields
> # you can also write a file name prefixed with "@" to include names
> # from a separate file.
> #
> # ADDRESS specifies the set of hosts the record matches.  It can be a
> # host name, or it is made up of an IP address and a CIDR mask that is
> # an integer (between 0 and 32 (IPv4) or 128 (IPv6) inclusive) that
> # specifies the number of significant bits in the mask.  A host name
> # that starts with a dot (.) matches a suffix of the actual host name.
> # Alternatively, you can write an IP address and netmask in separate
> # columns to specify the set of hosts.  Instead of a CIDR-address, you
> # can write "samehost" to match any of the server's own IP addresses,
> # or "samenet" to match any address in any subnet that the server is
> # directly connected to.
> #
> # METHOD can be "trust", "reject", "md5", "password", "scram-sha-256",
> # "gss", "sspi", "ident", "peer", "pam", "ldap", "radius" or "cert".
> # Note that "password" sends passwords in clear text; "md5" or
> # "scram-sha-256" are preferred since they send encrypted passwords.
> #
> # OPTIONS are a set of options for the authentication in the format
> # NAME=VALUE.  The available options depend on the different
> # authentication methods -- refer to the "Client Authentication"
> # section in the documentation for a list of which options are
> # available for which authentication methods.
> #
> # Database and user names containing spaces, commas, quotes and other
> # special characters must be quoted.  Quoting one of the keywords
> # "all", "sameuser", "samerole" or "replication" makes the name lose
> # its special character, and just match a database or username with
> # that name.
> #
> # This file is read on server startup and when the server receives a
> # SIGHUP signal.  If you edit the file on a running system, you have to
> # SIGHUP the server for the changes to take effect, run "pg_ctl reload",
> # or execute "SELECT pg_reload_conf()".
> #
> # Put your actual configuration here
> # ----------------------------------
> #
> # If you want to allow non-local connections, you need to add more
> # "host" records.  In that case you will also need to make PostgreSQL
> # listen on a non-local interface via the listen_addresses
> # configuration parameter, or via the -i or -h command line switches.
> 
> 
> 
> 
> # DO NOT DISABLE!
> # If you change this first entry you will need to make sure that the
> # database superuser can access the database using some other method.
> # Noninteractive access to all databases is required during automatic
> # maintenance (custom daily cronjobs, replication, and similar tasks).
> #
> # Database administrative login by Unix domain socket
> local   all             postgres                                peer
> 
> # TYPE  DATABASE        USER            ADDRESS                 METHOD
> 
> # "local" is for Unix domain socket connections only
> local   all             all                                     trust
> # IPv4 local connections:
> host    all             all             127.0.0.1/32            trust
> # IPv6 local connections:
> host    all             all             ::1/128                 trust
> # Allow replication connections from localhost, by a user with the
> # replication privilege.
> local   replication     all                                     trust
> host    replication     all             127.0.0.1/32            trust
> host    replication     all             ::1/128                 trust
> host    pretix          pretix          172.17.0.1/12           md5
> host    all             all             172.17.0.2/16           md5

**pretix.conf**

> [pretix]
> instance_name=My pretix installation
> url=https://tickets.musicalklasse.de
> currency=EUR
> ; DO NOT change the following value, it has to be set to the location of the
> ; directory *inside* the docker container
> datadir=/data
> trust_x_forwarded_for=on
> trust_x_forwarded_proto=on
> 
> [database]
> backend=postgresql
> name=pretix
> user=pretix
> ; Replace with the password you chose above
> password=
> ; In most docker setups, 172.17.0.1 is the address of the docker host. Adjust
> ; this to wherever your database is running, e.g. the name of a linked container.
> host=172.17.0.1
> 
> [mail]
> ; See config file documentation for more options
> from=
> ; This is the default IP address of your docker host in docker's virtual
> ; network. Make sure postfix listens on this address.
> host=172.17.0.1
> 
> [redis]
> location=unix:///var/run/redis/redis.sock?db=0
> ; Remove the following line if you are unsure about your redis' security
> ; to reduce impact if redis gets compromised.
> sessions=true
> 
> [celery]
> backend=redis+socket:///var/run/redis/redis.sock?virtual_host=1
> broker=redis+socket:///var/run/redis/redis.sock?virtual_host=2

I am using ufw:
Status: active

To                         Action      From
--                         ------      ----
22/tcp                     ALLOW       Anywhere
80/tcp                     ALLOW       Anywhere
443/tcp                    ALLOW       Anywhere
25/tcp                     ALLOW       Anywhere
Nginx Full                 ALLOW       Anywhere
8345/tcp                   ALLOW       Anywhere
5432/tcp                   ALLOW       Anywhere
22/tcp (v6)                ALLOW       Anywhere (v6)
80/tcp (v6)                ALLOW       Anywhere (v6)
443/tcp (v6)               ALLOW       Anywhere (v6)
25/tcp (v6)                ALLOW       Anywhere (v6)
Nginx Full (v6)            ALLOW       Anywhere (v6)
8345/tcp (v6)              ALLOW       Anywhere (v6)
5432/tcp (v6)              ALLOW       Anywhere (v6)

Any help would be highly appreciated.

**Comment by Nebelfuchs** [comment, marked as answer]
I solved it by moving my database into a container and used the docker bridge network

**Comment by Zorgosto** [comment]
I had the same error, was using the provided config in the docs so not sure if that needs correction or maybe more information.

Fixed it as well with putting both in a container and using a network.
If anyone is interested here is my pretix db config and docker-compose:

> [database]
backend=postgresql
name=<db_name>
user=<db_user>
; Replace with the password you chose above
password=<db_pw>
; In most docker setups, 172.17.0.1 is the address of the docker host. Adjust
; this to wherever your database is running, e.g. the name of a linked container.
; use docker compose service name of db
host=database

> services:
  database:
    image: postgres:15-bullseye
    container_name: database
    ports:
      - "5432:5432"
    environment: 
      POSTGRES_USER: <db_user>
      POSTGRES_PASSWORD: <db_pw>
      PGDATA: /data/postgres
    volumes:
      - ./data/volumes/postgres_data:/data/postgres
    networks:
      - backend
  pretix:
    image: pretix/standalone:2024.9.0
    container_name: pretix
    depends_on:
      - database
    ports:
      - "8345:80"
    volumes:
      - /var/pretix-data:/data
      - /etc/pretix:/etc/pretix
      - /var/run/redis:/var/run/redis
    networks:
      - backend
    sysctls:
      - net.core.somaxconn=4096

networks:
  backend:

I copied the data over with pg_dump using a temporarily different port for the docker postgres db:
`pg_dump -C -h localhost -U db_user db_name | psql -h localhost -p 5431 -U db_user db_name`

Of course you might have to change data path for volumes and use the correct postgresql version of your current db.

**Comment by F2b59** [comment]
The official documentation says to set `listen_addresses` in the file `/etc/postgresql/<version>/main/postgresql.conf`, so that it only listens on localhost and the docker gateway:
`listen_addresses = 'localhost,172.17.0.1'`
In my case, Pretix started working again after modifying it to:
`listen_addresses = '*'`

This certainly has security implications, but there may be a less radical solution than listening on '*'.
At least it's an alternative to avoid redeploying Postgres in a container.
